### PR TITLE
Feature/53/kademlia

### PR DESCRIPTION
**Pull request by Prodicode**
Added kademlia packets and packet handling.
Added kademlia packet forwarding.
Type of implemented kademlia packets up to this comment:
- ping
- ping_reply
- heartbeat

### CODE REVIEW

File: apps/purple/lib/purple/network/pipeline/packet_handler.ex [hunk 1]
```diff
@@ -176,6 +180,54 @@ defmodule Purple.Network.PacketHandler do
...
+        case Kademlia.unwrap_packet(kad_packet) do
+          {:ping, _} ->
+            Logger.debug("Got ping packet from #{Crypto.encode_base58(source)}. Sending reply!")
+            reply_packet = Kademlia.make_packet(:ping_reply, %{})
+            :ok = Kademlia.send(source, reply_packet)
+          {:ping_reply, _} ->
+            Logger.debug("Got ping reply from #{Crypto.encode_base58(source)}!");
+          {:heartbeat, event} ->
+            Logger.debug("Executing Heartbeat!")
+            {:ok, node} = PeersTrie.get(event.node_id)
```
Comment (octavonce) [inline apps/purple/lib/purple/network/pipeline/packet_handler.ex]: There is no alias for PeersTrie. Please add an alias for it.

File: apps/purple/lib/purple/network/kademlia.ex [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+  do
+    Logger.debug("Sending heartbeat to all sequencers!")
+    keys = PeersTrie.keys
+
+    case keys do
+      [] ->
+        {:error, :no_peers}
+      _ ->
+        _ = keys
+        |> Enum.each(fn(x) ->
```
Comment (octavonce) [inline apps/purple/lib/purple/network/kademlia.ex]: Please filter our node id so the packet won't get forwarded to ourselves.

**Comment by Prodicode** [comment]
Also send a heartbeat with kademlia in State Writer

File: apps/purple/lib/purple/network/pipeline/packet_decoder.ex [hunk 3]
```diff
@@ -150,6 +155,45 @@ defmodule Purple.Network.PacketDecoder do
...
+      packet_len   :: size(32),
+      kad_packet   :: binary-size(packet_len),
+      sign_len     :: size(32),
+      signature    :: binary-size(sign_len)
+    >>
+  }) do
+    Logger.debug("Received Kademlia packet! #{Crypto.encode_base58(source)} --> #{Crypto.encode_base58(destination)}")
+    {:ok, node_id} = NodeStorage.get("node_id")
+
+    with true <- Crypto.verify(signature, kad_packet, source) do
```
Comment (octavonce) [inline apps/purple/lib/purple/network/pipeline/packet_decoder.ex]: Why use with? Just do ```if Crypto.verify(signature, kad_packet, source) do```

File: apps/purple/lib/purple/network/pipeline/packet_handler.ex [hunk 4]
```diff
@@ -176,6 +184,62 @@ defmodule Purple.Network.PacketHandler do
...
+    peer_id: _,
+    packet: %{
+      type: :kad_heartbeat,
+      source: source,
+      destination: final_destination,
+      content: event,
+      signature: signature
+    }
+  }) do
+    with {:ok, event} <- Heartbeat.deserialize(event),
```
Comment (octavonce) [inline apps/purple/lib/purple/network/pipeline/packet_handler.ex]: Deserialization and validation should not happen in handler.

**Pull request merged**